[#15] Get extra files

GuChad369 opened the pull request:
This is just one step to start. I found that this function involves a lot of things and is not easy to achieve. I started this PR to help us discuss the details.

I want to describe my idea about this first.

I have created a function called `pgmoneta_ext_get_file(arg)` which returns the bytes of the file specified by the argument passed to it.

For the extension part, I prefer to create the following functions:
- `pgmoneta_ext_get_file(arg)`: This is the current version and it will return the bytes of the specified file passed in.
- `pgmoneta_ext_get_files(arg)`: This will return all file paths in the specified directory passed in.

For the core part, the process to get extra files is as follows:
- If we want to get a single file, we call `pgmoneta_ext_get_file` to get the bytes of the file and then pass it to the backend.
- If we want to get all files in a directory, we first use `pgmoneta_ext_get_files` to get all files in that directory. Then, we use a loop to pass all the files to the backend.

Achieving this function is not easy, and we may need to change it as we progress.

Commit:
[#15] Get extra files

GuChad369 commented:
How should we decide the privilege of getting the file? Currently, we only allow the `superuser` to get files.

jesperpedersen commented:
Well, PostgreSQL runs under an user account, so we should be able to use those credentials

jesperpedersen commented:
Try and follow the

```sql
SELECT session_user, current_user;
```

path...

GuChad369 commented:
Since `pgmoneta` will use the `repl` role to access PostgreSQL, should we give the privilege to the `repl` role?

Or do we need to define a function to grant a role the privilege to access the file?

`superuser` can use this function to grant the privilege to one user.

jesperpedersen commented:
There need to be options to for the `repl` role, and those will enable features in the extension.

So, a lot will be based on documentation for this

jesperpedersen commented:
We can start with a `SUPERUSER` requirement, and then improve over time

jesperpedersen commented:
So, from pgmoneta/core get the priviledge, and if it isn't `SUPERUSER` (to begin with), then you will log `ERROR` instead

GuChad369 commented:
Add a new step to cast binary data to characters to ensure we can handle this result on the core side. This will allow us to manage both text and binary files effectively.

GuChad369 commented:
These two functions can include a new feature to limit access to directories. We can add an array to list all directories that users can access, while other directories will not be accessible.

Alternatively, we can modify this later, incorporating user role limitations as well.

jesperpedersen commented:
I'm not following - we should just try to retrieve the files requested. If it doesn't work then `NULL` should be returned.

And, we don't want to scan the entire machine for directories that can be accessed...

GuChad369 commented:
> I'm not following - we should just try to retrieve the files requested. If it doesn't work then `NULL` should be returned.

This is the current behavior of the function.

> And, we don't want to scan the entire machine for directories that can be accessed...

Since I saw requests for access outside the `/data` directory, I meant we can add a step to check if the file path is under `/data`. If it is, then return `NULL`.

jesperpedersen commented:
I don't think you need to do that - maybe the user want an extra copy of `pg_control`

jesperpedersen commented on src/include/utils.h:
```diff
@@ -70,6 +76,14 @@ pgmoneta_ext_check_role(Oid roleid, const char* rolename);
+ * @param astate The array build state to accumulate file paths.
+ */
+void
+list_files(const char* name, ArrayBuildState* astate);
```
Please, prefix with `pgmoneta_ext_` to all functions in `.h` files

jesperpedersen replied:
Should this be `static` ?

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -28,31 +28,53 @@
 /* pgmoneta_ext */
-#include <pgmoneta_ext.h>
-#include <utils.h>
+#include "pgmoneta_ext.h"
```
Why the `"` ?

jesperpedersen commented on src/pgmoneta_ext/utils.c:
```diff
@@ -28,18 +28,18 @@
 
 /* pgmoneta_ext */
-#include <utils.h>
+#include "utils.h"
```
Again, `"`

jesperpedersen commented:
This doesn't compile with PostgreSQL 16 - test with 13->17, and add the functions to the manual

jesperpedersen commented:
```
In file included from pgmoneta/pgmoneta_ext/src/include/utils.h:39,
                 from pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c:32:
pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c: In function ‘pgmoneta_ext_get_files’:
pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c:305:29: error: passing argument 1 of ‘PointerGetDatum’ makes pointer from integer without a cast [-Wint-conversion]
  305 |       PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate, CurrentMemoryContext));
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                             |
      |                             Datum {aka long unsigned int}
/usr/local/packages/postgresql-16.x/include/server/fmgr.h:361:53: note: in definition of macro ‘PG_RETURN_POINTER’
  361 | #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
      |                                                     ^
pgmoneta/pgmoneta_ext/src/pgmoneta_ext/lib.c:305:7: note: in expansion of macro ‘PG_RETURN_ARRAYTYPE_P’
  305 |       PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate, CurrentMemoryContext));
      |       ^~~~~~~~~~~~~~~~~~~~~
In file included from pgmoneta/pgmoneta_ext/src/include/utils.h:38:
/usr/local/packages/postgresql-16.x/include/server/postgres.h:322:29: note: expected ‘const void *’ but argument is of type ‘Datum’ {aka ‘long unsigned int’}
  322 | PointerGetDatum(const void *X)
      |                 ~~~~~~~~~~~~^
make[2]: *** [src/CMakeFiles/pgmoneta_ext.dir/build.make:76: src/CMakeFiles/pgmoneta_ext.dir/pgmoneta_ext/lib.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:935: src/CMakeFiles/pgmoneta_ext.dir/all] Error 2
make: *** [Makefile:166: all] Error 2
```

Maybe we need changes to test suite...

Compiled with

```
export CC=/usr/bin/gcc && cmake -DCMAKE_BUILD_TYPE=Debug ..
```

but we may need

```
export CC=/usr/bin/clang && cmake -DCMAKE_BUILD_TYPE=Debug ..
```

as well

jesperpedersen commented:
Why did you disable the CI ?

jesperpedersen commented:
Make sure that you are happy with the patch as a whole, and then send a PTAL

GuChad369 commented:
> Why did you disable the CI?

To make testing more convenient. To test our test suite in a container, I need to commit changes to enable the container to download the updated code version. Therefore, I disabled the CI to speed up the testing process.

I noticed it is available for version 17, so I have added PostgreSQL version 17 to our test suite, and now everything works well in the container.

I have also added new tests for our new functions.

![Screenshot 2024-08-06 at 10 19 34 AM](https://github.com/user-attachments/assets/04654652-573f-4be7-9ccc-3307e186ea02)

There is one issue with the CI: our test suite needs to retrieve a file from the container, which does not exist in the CI environment, causing it to return an error. How should we handle this in the CI?

![Screenshot 2024-08-06 at 10 56 11 AM](https://github.com/user-attachments/assets/c6530cbb-e00a-4c59-b8cc-88dcb5b50bb1)

GuChad369 commented:
Can we remove the test suite from the CI and only call a single function, like `pgmoneta_ext_version`, to ensure it installs successfully?

jesperpedersen commented:
We can disable the test suite in CI, check if there something that need to be updated for the developer guide

jesperpedersen commented on doc/manual/user-02-functions.md:
```diff
@@ -13,5 +13,10 @@ After you create the extension `pgmoneta_ext` using the `postgres` role, you can
 | `pgmoneta_ext_version()`    |   Default |    None    | Return the version number of `pgmoneta_ext` as a Datum.|
 | `pgmoneta_ext_switch_wal()` | SUPERUSER |    None    | A function for switching to a new WAL file.            |
 | `pgmoneta_ext_checkpoint()` | SUPERUSER <br>pg_checkpoint | None | A function which forces a checkpoint. <br>This function can only be executed by a `SUPERUSER` in PostgreSQL 13/14, but can also be executed by `pg_checkpoint` in PostgreSQL 15+. <br>You can use the SQL command `GRANT pg_checkpoint TO repl;` to assign the role in PostgreSQL 15+.|
+<<<<<<< HEAD
```
Whoops

GuChad369 commented:
This new version is all set and can work on all versions from 13 to 17.

If you want to test, please remember to replace line 80 in Dockerfile.rocky9 with:
``` Docker
RUN git clone --branch "extra" --single-branch --depth 1 https://github.com/GuChad369/pgmoneta_ext
```

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -195,6 +218,7 @@ pgmoneta_ext_get_oid(PG_FUNCTION_ARGS)
    }
 }
 
+
```
Extra linefeed

jesperpedersen commented on src/include/utils.h:
```diff
@@ -38,13 +38,19 @@ extern "C" {
 /* system */
+#include <dirent.h>
 #include <stdbool.h>
+#include <sys/stat.h>
```
Direct include first, then directory includes after - all sorted

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -38,23 +38,46 @@
+#include <sys/stat.h>
+#include <unistd.h>
+#include <dirent.h>
+
```
Sorting

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -38,23 +38,46 @@
+#include <unistd.h>
+#include <dirent.h>
+
+#ifdef PG_MODULE_MAGIC
```
Why `#ifdef` ?

GuChad369 replied:
This got messed up by the rebase.

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -255,3 +279,178 @@ pgmoneta_ext_get_oids(PG_FUNCTION_ARGS)
+         PG_RETURN_NULL();
+      }
+
+      // PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate, CurrentMemoryContext));
```
Comment

GuChad369 replied:
The reason for the version 16 error is a type mismatch. I changed this using a simple method and will retest it.

jesperpedersen commented on src/pgmoneta_ext/utils.c:
```diff
@@ -35,8 +35,8 @@
 #include <catalog/indexing.h>
-#include <catalog/pg_authid.h>
 #include <catalog/pg_database.h>
+#include <catalog/pg_authid.h>
```
Sorting

GuChad369 commented:
This new version works well.

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -38,23 +38,44 @@
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <dirent.h>
```
Sorting

jesperpedersen replied:
Sorting

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -255,3 +276,176 @@ pgmoneta_ext_get_oids(PG_FUNCTION_ARGS)
+   }
+   else
+   {
+      ereport(LOG, errmsg_internal("pgmoneta_ext_checkpoint: Current role is not a superuser"));
```
`_get_file`

jesperpedersen replied:
Still `_checkpoint`

jesperpedersen commented on src/pgmoneta_ext/lib.c:
```diff
@@ -255,3 +276,176 @@ pgmoneta_ext_get_oids(PG_FUNCTION_ARGS)
+   }
+   else
+   {
+      ereport(LOG, errmsg_internal("pgmoneta_ext_checkpoint: Current role is not a superuser"));
```
`_get_files`

jesperpedersen replied:
Again

jesperpedersen commented:
Look at your patch, and ask yourself if every single change is correct.

It will make for much faster merges

GuChad369 commented:
This is the newly updated version.

jesperpedersen commented:
Merged.

Thanks for your contribution !
